jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -719,38 +720,104 @@ protected void acquire()
         }
     }
 
+    /**
+     * A variant of the {@link ArrayByteBufferPool} that
+     * uses a predefined set of buckets of buffers.
+     */
+    public static class Predefined extends ArrayByteBufferPool
+    {
+        public Predefined(int... capacities)
+        {
+            this(0L, 0L, capacities);
+        }
+
+        public Predefined(long maxHeapMemory, long maxDirectMemory, int... capacities)
+        {
+            super(sort(capacities)[0], 1, capacities[capacities.length - 1], Integer.MAX_VALUE, maxHeapMemory, maxDirectMemory,
+                c -> indexOfClosestBelow(c, capacities), i -> capacityOfIndex(i, capacities));
+        }
+
+        private static int[] sort(int... values)
+        {
+            if (values.length == 0)
+                throw new IllegalArgumentException("At least one capacity is needed");
+            Arrays.sort(values);
+            return values;
+        }
+
+        private static int capacityOfIndex(int idx, int... capacities)
+        {
+            if (idx >= capacities.length)
+            {
+                int largestCapacity = capacities[capacities.length - 1];
+                return (idx - capacities.length + 2) * largestCapacity;
+            }
+            return capacities[idx];
+        }
+
+        private static int indexOfClosestBelow(int capacity, int... capacities)
+        {
+            int largestCapacity = capacities[capacities.length - 1];
+            if (capacity > largestCapacity)
+            {
+                int remainder = capacity % largestCapacity != 0 ? 1 : 0;
+                return capacity / largestCapacity - 1 + remainder + capacities.length - 1;
```

Comment:
Needs description of what the logic is.

Given `[128, 512, 2048]`, for `8192` it returns `6`, which fed to the other function yields `10240`.
Is that wanted? If so, needs to explain why.

Reply:
This is only used to make `recordNoBucketAcquire()` record somewhat meaningful capacities. I've added some comments to explain that.